core/src/main/java/org/apache/calcite/rel/rules/IntersectToDistinctRule.java

```diff
@@ -96,39 +91,48 @@ public IntersectToDistinctRule(Class<? extends Intersect> intersectClass,
     final RelOptCluster cluster = intersect.getCluster();
     final RexBuilder rexBuilder = cluster.getRexBuilder();
     final RelBuilder relBuilder = call.builder();
-
-    // 1st level GB: create a GB (col0, col1, count() as c) for each branch
-    for (RelNode input : intersect.getInputs()) {
-      relBuilder.push(input);
-      relBuilder.aggregate(relBuilder.groupKey(relBuilder.fields()),
-          relBuilder.countStar(null));
+    final String countPrefix = "count_i";
+    final String indexAlias = "i";
+    final int oriFieldCount = intersect.getRowType().getFieldCount();
+    final int inputCount = intersect.getInputs().size();
+
+    AggCall[] aggCalls = new AggCall[inputCount];
+    for (int i = 0; i < inputCount; ++i) {
+      relBuilder.push(intersect.getInputs().get(i));
+      List<RexNode> fields = new ArrayList<>(relBuilder.fields());
+      // Add a constant column "i" for each input
+      fields.add(
+          relBuilder.alias(
+              rexBuilder.makeBigintLiteral(new BigDecimal(i)), indexAlias));
+      relBuilder.project(fields);
+      // Generate count(*) filter (where i = n) as count_i{n}
+      aggCalls[i] = relBuilder.countStar(null)
+          .filter(
+              relBuilder.equals(relBuilder.field(indexAlias),
+              relBuilder.literal(i))).as(countPrefix + i);
     }
 
     // create a union above all the branches
-    final int branchCount = intersect.getInputs().size();
-    relBuilder.union(true, branchCount);
+    relBuilder.union(true, inputCount);
     final RelNode union = relBuilder.peek();
-
-    // 2nd level GB: create a GB (col0, col1, count(c)) for each branch
-    // the index of c is union.getRowType().getFieldList().size() - 1
     final int fieldCount = union.getRowType().getFieldCount();
 
-    final ImmutableBitSet groupSet =
-        ImmutableBitSet.range(fieldCount - 1);
-    relBuilder.aggregate(relBuilder.groupKey(groupSet),
-        relBuilder.countStar(null));
+    // Add aggCalls
+    relBuilder.aggregate(relBuilder.groupKey(ImmutableBitSet.range(fieldCount)), aggCalls);
+
+    // Generate filter count_i{n} > 0 for each branch
+    RexNode[] filters = new RexNode[inputCount];
+    for (int i = 0; i < inputCount; i++) {
+      filters[i] =
+          relBuilder.greaterThan(relBuilder.field(countPrefix + i), relBuilder.literal(0));
+    }
 
-    // add a filter count(c) = #branches
-    relBuilder.filter(
-        relBuilder.equals(relBuilder.field(fieldCount - 1),
-            rexBuilder.makeBigintLiteral(new BigDecimal(branchCount))));
+    // Add filter count_i{n} > 0 for each branch
+    relBuilder.filter(relBuilder.and(filters));
```

Comment:
`relBuilder.filter(relBuilder.and(filters))` &rarr; `relBuilder.filter(filters)`